src/oop/object_interactions/medium/q2/MainTest.java

```diff
@@ -0,0 +1,218 @@
+                Arguments.of("John", "John is in love with Alice", new String[]{"Alice", "Bob", "Sam", "Neil", "Jack", "Tom", "Sara"}, "John", "Alice is in love with John"),
+                Arguments.of("Kim", "Kim has broken her leg", new String[]{"Ali", "John", "Alice", "Bob", "Sam", "Neil", "Jack", "Tom", "Sara"}, "Ali", "Ali broke Kim's leg")
+
+        );
+    }
+
+    @Test
+    public void hearSecretIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}), String.format(incorrectMethodDefinition, hearSecretMethodName, className));
+        assertTrue(testClass.hasModifier(hearSecretMethodName, new Class<?>[]{String.class, String.class}, "public"), String.format(incorrectModifierMessage, hearSecretMethodName, className));
+        assertTrue(testClass.hasReturnType(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE), String.format(incorrectReturnType, hearSecretMethodName, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("hearSecretInputProvider")
+    @Tag("dependent1")
+    public void correctHearSecretMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {secret, String.class},
+                {allFriends, String[].class}
+        };
+        Object[][] methodArguments = {
+                {friend, String.class},
+                {friendSecret, String.class}
+        };
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}),
+                "Your " + hearSecretMethodName + " method does not have the correct name or arguments.");
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE),
+                "Your " + hearSecretMethodName + " method does not have the correct return type.");
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE, new String[]{"public"}),
+                "Your " + hearSecretMethodName + " method does not have the correct modifier.");
+        Object classInstance = testClass.createInstance(arguments);
+        Object output = testClass.callMethod(hearSecretMethodName, methodArguments, new String[]{"public"}, classInstance);
+        String[] ans = new String[allFriends.length];
+        for (int i = 0; i < allFriends.length; i++) {
```

Comment:
This should probably change, because here we pass them the friend's array, which probably does not get cloned in the constructor. If they modify it at all, it could cause our test to break because it will modify the array we passed them.

Reply:
Also, are you testing if it is not a valid name?

Reply:
I think you could just pass the index where to input the secret or something, which would also make this test easier.